### DISCUSSION TITLE
Duplicate G7 CGM Sensor Insert

### DISCUSSION THREAD

**Discussion by Knighth4wk** [NightscoutFoundation/xDrip#3508] [Bug reports]
Hello,

I am using xDrip+AAPS+NS setup. I recently updated to version 20240419 of xDrip and noticed mutiple duplicate entries every hour for CGM sensor start in my NS and AAPS treatments. AAPS is setup to receive BG data from xDrip and upload to NS.

I went through previous treatments entries and confirmed the duplicates started showing up right after updating xDrip to latest stable release.

Current xDrip Version: 20240419
Previous xDrip Version: 20230716
xDrip Data Source: G7 sensor
AAPS version: 3.2.0.4
Nightscout: 15.0.2
Phone: Samsung Galaxy S24 Ultra running Android 14
![Screenshot_20240601_150353_xDrip+](https://github.com/NightscoutFoundation/xDrip/assets/40640092/6acb81db-f526-4f1e-98f3-f981c5cd7b82)
![Screenshot_20240601_150403_xDrip+](https://github.com/NightscoutFoundation/xDrip/assets/40640092/944a5c40-38ef-455f-91a3-4713619c4da0)
![Screenshot_20240531_185807_AAPS](https://github.com/NightscoutFoundation/xDrip/assets/40640092/b7ff01a6-33c3-48c8-b570-1b0fb83d62fb)

**Comment by Navid200** [comment]
xDrip has changed significantly for better performance with G7.  You should use a more recent release as suggested in the guide: https://navid200.github.io/xDrip/docs/Dexcom/G7.html

Are you only uploading to Nightscout from AAPS and not also from xDrip?

**Reply by Knighth4wk** [reply]
I will try out the May 28 release and report back.

And only AAPS is uploading to Nightscout.

Thanks

**Reply by Navid200** [reply]
xDrip had a bug that affected AAPS.  It has been fixed.
I hope this is related and will be fixed after you update.
If not, please tell us.

**Reply by Knighth4wk** [reply]
Updating to May 28 release did not fix the issue.

**Reply by Navid200** [reply]
@olorinmaia is quite knowledgeable about AAPS and G7 and xDrip.
Hopefully, he can shed some light on this.

**Reply by olorinmaia** [reply]
I've checked my son's AAPS and I notice the duplicate entries are there also. There are no duplicates in Nightscout however, therefor no duplicates in AAPS follow app, AAPSClient which is the app I mostly browse treatments history in, that's why it has been hard for me to discover this. Or if this has been happening earlier.

@Knighth4wk you are sure this didn't happen before 20140419 release? If so it is probably this commit https://github.com/NightscoutFoundation/xDrip/commit/1d77e78019f6291fd23063a9abcc2d38a2f039ed that has changed something within AAPS how the broadcast information is received. I've asked Milos for advice, I'll let you know what he says.

No duplicates in NS:
![image](https://github.com/NightscoutFoundation/xDrip/assets/114103483/530bf3b3-ae83-4d45-a986-39b706fe3936)
Duplicates in Treatments log > Careportal and User entry (They seem to be added every 5 min or recieved BG?):


| User entry | Careportal |
|--------|--------|
| ![image](https://github.com/NightscoutFoundation/xDrip/assets/114103483/2ae4a6da-5377-4849-a9ed-fe43614bc41a) | ![image](https://github.com/NightscoutFoundation/xDrip/assets/114103483/f86511a7-5b97-4210-b8f8-7207c8058cf7) |

**Reply by Knighth4wk** [reply]
I have double checked NS, the duplicated entries are showing there too for me.
![image](https://github.com/NightscoutFoundation/xDrip/assets/40640092/144c6412-e3e8-42cb-ba77-416569957805)

The previous version I was using 20230716 did not have this issue.

**Reply by olorinmaia** [reply]
Are you using NSClientV1 or NSClientV3?

**Reply by Knighth4wk** [reply]
I am using NSClientV3.

**Reply by olorinmaia** [reply]
NSClientV1 here. That could be the reason for why you get duplicates in NS and I don't. It's very hard for me to see in the logs why this is happening in AAPS, need help from @MilosKozak or someone more experienced. Here are logs from AAPS in a period this issue is occuring:
[AndroidAPS_LOG_1717346137057.log.zip](https://github.com/user-attachments/files/15540287/AndroidAPS_LOG_1717346137057.log.zip)

**Comment by MilosKozak** [comment]
```
18:29:32.658 [DefaultDispatcher-worker-7] D/BGSOURCE: [XdripSourcePlugin$XdripSourceWorker.doWorkAndLog():102]: Received xDrip data: Bundle[{com.eveningoutpost.dexdrip.Extras.NoiseBlockLevel=200, com.eveningoutpost.dexdrip.Extras.BgSlope=7.348103020404346E-5, com.eveningoutpost.dexdrip.Extras.VersionInfo=2124052808 c67b74f-2024.05.28, com.eveningoutpost.dexdrip.Extras.Raw=0.0, com.eveningoutpost.dexdrip.Extras.Display.Units=mmol, com.eveningoutpost.dexdrip.Extras.Time=1717345763598, com.eveningoutpost.dexdrip.Extras.CalibrationInfo=-1, com.eveningoutpost.dexdrip.Extras.NsNoiseLevel=null, com.eveningoutpost.dexdrip.Extras.SensorStartedAt=1716739867561, com.eveningoutpost.dexdrip.Extras.Noise=-9999.0, com.eveningoutpost.dexdrip.Extras.SensorBattery=-1, com.eveningoutpost.dexdrip.Extras.BgSlopeName=FortyFiveUp, com.eveningoutpost.dexdrip.Extras.BgEstimate=150.0, com.eveningoutpost.dexdrip.Extras.SourceDesc=G6 Native / G5 Native, com.eveningoutpost.dexdrip.Extras.SourceInfo=G7, com.eveningoutpost.dexdrip.Extras.Collector.NanoStatus=  }]
18:34:33.391 [main] D/CORE: [DataReceiver.onReceive():35]: onReceive com.eveningoutpost.dexdrip.BgEstimate Bundle{ com.eveningoutpost.dexdrip.Extras.NoiseBlockLevel => 200; com.eveningoutpost.dexdrip.Extras.BgSlope => 0.0; com.eveningoutpost.dexdrip.Extras.VersionInfo => 2124052808 c67b74f-2024.05.28; com.eveningoutpost.dexdrip.Extras.Raw => 0.0; com.eveningoutpost.dexdrip.Extras.Display.Units => mmol; com.eveningoutpost.dexdrip.Extras.Time => 1717346064332; com.eveningoutpost.dexdrip.Extras.CalibrationInfo => -1; com.eveningoutpost.dexdrip.Extras.NsNoiseLevel => null; com.eveningoutpost.dexdrip.Extras.SensorStartedAt => 1716739868294; com.eveningoutpost.dexdrip.Extras.Noise => -9999.0; com.eveningoutpost.dexdrip.Extras.SensorBattery => -1; com.eveningoutpost.dexdrip.Extras.BgSlopeName => FortyFiveUp; com.eveningoutpost.dexdrip.Extras.BgEstimate => 150.0; com.eveningoutpost.dexdrip.Extras.SourceDesc => G6 Native / G5 Native; com.eveningoutpost.dexdrip.Extras.SourceInfo => G7; com.eveningoutpost.dexdrip.Extras.Collector.NanoStatus =>   ; }Bundle
```
Sensor start is received multiple times with different timestamp.
AAPS does deduplication on exact timestamp only

Here:
Sun May 26 2024 18:11:08 GMT+0200 (středoevropský letní čas)
Sun May 26 2024 18:11:07 GMT+0200 (středoevropský letní čas)

**Reply by Navid200** [reply]
I understand that we have a problem.
Before we can fix it, I need to exactly understand what the problem is.

Is xDrip getting a timestamp from Dexcom, but broadcasts a different timestamp to AAPS?
Is xDrip receiving a timestamp and broadcasts it exactly as is to AAPS, but regardless, ....?

I don't use AAPS.  Therefore, I cannot see the problem as well as one who uses AAPS.
When I look at my G7 timestamps in xDrip, I cannot see anything wrong.
But, I am not surprised because I don't use AAPS. 

Can you help me see exactly what the problem is?
Then, I hope to be able to recreate the problem for myself, identify the cause, and fix it, or ask someone else to fix it if I cannot.

**Reply by Navid200** [reply]
A Dexcom device has a free running local oscillator.  The timebase of the device is based on this oscillator.
The oscillator is never synchronized with the internet.  That's what I mean by free running.
The result is that the timebase may not be perfect.  It is quite possible that the time will gradually drift.

I started my current G7 3.7 days ago.  Today is June 5.
Please look at the following timestamps I am going to show from the past few days all from the same G7.

![Screenshot_20240605-231335](https://github.com/NightscoutFoundation/xDrip/assets/51497406/a2de9fb8-add2-4f7c-9229-4762005a67e5)
In the above screenshot, you can see readings from June 3.  And you can see 27 seconds for the seconds value of time.

The next image shows readings from June 4 and you can see that the seconds value is 28 seconds.  But, please note that one timestamp is actually 29 seconds!  I will talk about this next.  
![Screenshot_20240605-231359](https://github.com/NightscoutFoundation/xDrip/assets/51497406/c5ddef9b-a9a0-43db-8c4a-c45127278ded)

Please note that because the oscillator is free running, its frequency is not perfect.  As a result, it drifts.  The result is that its seconds value is either going to gradually increase or decrease.
But, the timebase of the Dexcom device is not an extremely sophisticated oscillator.  Therefore, it can have phase noise.  The result of the phase noise is that at every second boundary cross-over, there will be some back and forth.  And that's the reason for the single different seconds value in the previous set.

The next screenshot shows readings from June 5.  But, I intentionally picked a time period where you can clearly see the back and forth caused by phase noise.  You can see 29 and 30 and 29 and 30.  
![Screenshot_20240605-231444](https://github.com/NightscoutFoundation/xDrip/assets/51497406/d3510091-5da4-4da8-85f4-36f34c802397)

<br/>  
If I understand the issue you are talking about correctly and it being the time value changing, the reason is that the timebase of the device is not synchronized to the internet.

Please tell me what you think.

Thanks

**Reply by MilosKozak** [reply]
i think the problem is only with timestamp of sensor start , not the BG itself

**Reply by olorinmaia** [reply]
Here are three examples from log where SensorStartedAt timestamp varies:
05:54:31.143 [main] D/CORE: [DataReceiver.onReceive():35]: [......] com.eveningoutpost.dexdrip.Extras.SensorStartedAt => **1716739865948**;
05:49:30.619 [main] D/CORE: [DataReceiver.onReceive():35]: [......] com.eveningoutpost.dexdrip.Extras.SensorStartedAt => **1716739866450**;
05:44:34.874 [main] D/CORE: [DataReceiver.onReceive():35]: [......] com.eveningoutpost.dexdrip.Extras.SensorStartedAt => **1716739865780**;

1716739865948 => May 26, 2024, 16:11:05.948
1716739866450 => May 26, 2024, 16:11:06.450
1716739865780 => May 26, 2024, 16:11:05.780

Timestamp for G7 sensorstart is now fetched from transmitter itself as I understand it. But why does it vary for each broadcast that's a different question.

**Reply by Navid200** [reply]
I see,
Thanks
I had misunderstood the problem.

The transmitter never tells us when the sensor started because it never knows what time it is!
It only tells us how long ago the sensor started.  But, it knows how long ago the sensor started based on its local oscillator perception of time, which I already explained can be wrong.

It's xDrip that takes this relative time value and converts it into a real time.  But, xDrip uses the time of the phone to calculate that.

This is a change I made for G7.  So, I am responsible for this mess.  Sorry about that.
Let me go to the code in question and see how I can fix it.

Thanks for reporting this

**Reply by Navid200** [reply]
I cannot think of a solution!
@jamorham 

Unless someone maintains a database reporting and maintaining the time inaccuracy of G7 (and G6) transmitters, there is no way to convert started 2.5 days ago to started at .... without a drifting error.  

Let's go out of our way to fix this.  Let's say when we first establish connectivity to a device, we create a database that contains when exactly this device started.  Then, we can ignore the relative start time reported by the transmitter.
But, what happens if you install xDrip on a second phone and connect to the same device?  Now, xDrip on the second phone is going to calculate the start time to store in the xDrip database on the second phone.  But, this start time will be different than the one stored in the database of xDrip on the first phone.  
That's not going to work.

Why is a slight change to the device start time so critical for AAPS?

**Reply by MilosKozak** [reply]
The problem is records duplication because of different timestamps.
I think the solution could be round up the start time to a whole minute. It's sufficient precision for what we need

**Reply by Navid200** [reply]
I can do that.
That will effectively reduce the probability of change significantly.  But, it will not zero the probability.

What I mean is that if the start time happen to end up at a round-up boundary, there will still be fluctuations for a while at the cross over point.  
But, statistically speaking, your suggestion will reduce the probability by a factor of 60 * 1000.

**Reply by MilosKozak** [reply]
In logs the fluctuation is within a second. Thus I offered rounding to minutes but it can be anything else

**Reply by Navid200** [reply]
Let's say we insert the sensor now and lets say it is 9:00:00 now.

In exactly 5 days at 9:00:00, the transmitter says that the sensor started 5 days and 5 seconds ago instead of 5 days ago because of the drift of the local oscillator.
So, the calculated start time becomes 8:59:55.  And we round it and it will still become 9:00.  So, no problem.

But, what if we insert the sensor at 9:00:32?
We will round it to 9:01 and show that as the start time.
In 5 days when the transmitter says that the sensor started 5 days and 5 seconds ago, that translates to 9:00:27.  That rounds to 9:00.
So, the start time changes from 9:01 to 9:00. 

This is not going to be as bad as it is now where every millisecond we may report a different start time.
But, I want you to know that this is not a perfect solution.
Even if we round to hour, this could still happen.  But, the probability will improve compared to what we have now by a factor of 60*60*1000.
But, no matter what we do, the probability will never be 0.
If I'm not mistaken, Dexcom app only reports started 5.2 days ago.
Even that does not lower the probability to 0.

**Reply by Knighth4wk** [reply]
I am not familiar with the code at all. Just throwing an idea out there. If new time is within +/- 5 minutes of previous reported time, then use previous time. Else use new reported time?

Please ignore me if this does not make sense. I haven't touched a code in a long time.

**Reply by Navid200** [reply]
@Knighth4wk I mentioned the concept of using the history in xDrip as a judgement mechanism.  And I explained that going to a different xDrip, you may then get a different start time since the history remains with the previous xDrip and is not transferred to a new fresh install.

**Reply by Navid200** [reply]
Can AAPS be modified so that it does not trigger an alert if the sensor start time changes?

**Reply by Navid200** [reply]
If what matters to AAPS is the age of the sensor, AAPS can subtract the reported start time from the current time to come up with the sensor age.
This can be rounded to minutes and the result will never have any fluctuations.

Alternatively, we can add a new item to broadcast from xDrip that being the sensor age.  But, I am not sure if I can get approval for it.  I hope I can.
@jamorham

**Reply by Navid200** [reply]
The transmitter reports once every 5 minutes how long ago the session started.  It does this using its internal clock.  The error in the value that it reports is negligible (very small percentage).

The problem starts when we want to translate that into real (absolute) time.
Meaning instead of saying the session started 4.1 days ago, we say the session started on June 2 at 12:00:17.

Why do we need to do that?
Can you tell me what requirement exists to use absolute time in this case and why relative time will not work?

**Reply by Navid200** [reply]
Forgive me for post after post after post.  But, I am not sure if we have a path forward.  If every one of us assumes that the other person has to do something, this may never be resolved.

We have unintentionally created a vernier here: https://en.wikipedia.org/wiki/Vernier_scale

What the transmitter transmits is how long ago sensor started.  We, xDrip, subtract that from the current time to calculate when exactly the sensor started.  The accuracy of the result depends on the accuracy of both subtraction operands.  The problem is that the relative time coming from the transmitter is based on the local oscillator of the transmitter while the current time is based on internet time.  Those two time bases are drifting with respect to each other.
So, it is like someone is continuously sliding the vernier.  The vernier reading is continuously being changed because of the drift.

**Solution:**  
xDrip to broadcast the sensor age exactly as reported by the transmitter, rather than when it started.  I (Navid) to submit a PR.  
AAPS to use the incoming sensor age to do all its internal calculations.

@jamorham Is this acceptable to you?
@MilosKozak Would this be acceptable to you?

**Reply by olorinmaia** [reply]
@MilosKozak @jamorham @Navid200 Issue persist in AAPS 3.3. Given Navids explanations above and possible solutions are we closer to get this improved? I'm not sure what is the best approach but there are some suggestions here and perhaps the best solution hasn't surfaced yet either. 

See example of treatment log in AAPS 😅
https://photos.app.goo.gl/sNAxgjEXDs5o1VN37

**Reply by Navid200** [reply]
I can open a PR and add a new parameter to broadcast and that being how long ago from the transmitter's perspective the transmitter started.  This will still be the perception of the transmitter.
The problem of the transmitter not knowing exactly what time it is will never go away.
Either xDrip or AAPS needs to make a judgement call knowing the limitation of the transmitter and address the possible issue.

One way to address this is to have a record.  So, for every transmitter ID, there is a start time.  The first time we receive data from that transmitter telling us how long ago it started, we subtract that from current time and record the result as well as the transmitter ID.
After that every time the transmitter tells us how long ago it started, we just look at the records in the database.  If we find the transmitter in the database, we just ignore what it is telling us about when it started and use our initial calculation from the database.

This will address the time drift of the transmitter.
This can be done right now as is by AAPS with no change required to xDrip.  

I mean a transmitter start time should not change.  Even though xDrip reports a different start time for the same transmitter, AAPS can continue to use the first start time that xDrip reported for it and ignore the subsequent start times reported for the same transmitter.

**Reply by Navid200** [reply]
A long time ago, Dexcom had a product called G5.  Most of the intelligence was in the receiver/Dex app.
That was why it was so easy to restart a sensor.
xDrip was written even before then for G4 to do everything including calibration and sensor control.  So, if you used the xDrip's mechanism, the sensor would not atop after 7 days.
xDrip maintained a session that represented when the sensor started.  xDrip remembered a time not a relative time representing when the session started.  In case of the user using xDrip and not native mode, it did not matter when the session started on the transmitter because xDrip was in full control because there was not much intelligence in the transmitter.  And what we broadcast is the absolute time that does not change.

This changed significantly when G6 came along and its firmware was later changed.  Effectively, the entire intelligence moved to the transmitter.
For backwards compatibility, xDrip kept maintaining its local sensor for a G6 even though it did not have too much significance.  But, the code relied on it.

When G7 came along, nothing changed in that respect.  But, G7 offered a grace period.  Many of us insert a new G7 sensor while our old sensor still has 12 hours life left.
When the time comes to switch to the new sensor, it has actually been running for 12 hours.
This really messes up the concept of the xDrip internal sensor.  

We have to gradually move and update xDrip to focus on what matters and gradually retire what is there only for backwards compatibility to technology that no longer exists.
I am sorry that one of these changes has caused grief for you guys.  That has not been our intent.
We don't broadcast the start time of the internal sensor for G6 or G7 any longer.  
We still do have an internal sensor for G7.  But, it cannot have a valid start time because the real start time is 12 hours before the previous sensor's end time!
The G7 xDrip internal sensor is only there to let us use some of the existing code instead of reinventing the wheel.  
  
If you think that xDrip broadcasting a new parameter of relative session start time can help you, please tell me and I will open a PR in a day or two.  But, the underlying problem will never go away; the transmitter has no idea what time it is!

**Reply by olorinmaia** [reply]
@Navid200 I think this sounds like a good solution if I understand you right. 

Is the identifier for saving in xDrip DB the 4 digit code aka dexcom sender-ID? If so it should be cleared when entering a new 4 digit code on next sensor since I have seen these codes repeat themselves.

**Reply by Navid200** [reply]
Do you see a hex value similar to this in AAPS?
![403631315-31e5455d-06c7-4114-9d53-f0bba37b98f4](https://github.com/user-attachments/assets/cab40eb6-d31a-491b-a9bd-23971a4dd8b3)

**Reply by olorinmaia** [reply]
This is what is sent from xDrip to AAPS via local broadcast:

10:33:55.961 [main] D/CORE: [DataReceiver.onReceive():38]: onReceive com.eveningoutpost.dexdrip.BgEstimate Bundle{ com.eveningoutpost.dexdrip.Extras.NoiseBlockLevel => 200; com.eveningoutpost.dexdrip.Extras.BgSlope => -6.669378880744836E-6; com.eveningoutpost.dexdrip.Extras.VersionInfo => 2124102913 d14910c-2024.10.29; com.eveningoutpost.dexdrip.Extras.Raw => 0.0; com.eveningoutpost.dexdrip.Extras.Display.Units => mmol; com.eveningoutpost.dexdrip.Extras.Time => 1737797625902; com.eveningoutpost.dexdrip.Extras.CalibrationInfo => -1; com.eveningoutpost.dexdrip.Extras.NsNoiseLevel => null; 

**com.eveningoutpost.dexdrip.Extras.SensorStartedAt => 1737718830827;** 

**The timestamp 1737718830827 corresponds to January 24, 2025, 11:40:30.827 AM according to ChatGPT**

com.eveningoutpost.dexdrip.Extras.Noise => -9999.0; com.eveningoutpost.dexdrip.Extras.SensorBattery => -1; com.eveningoutpost.dexdrip.Extras.BgSlopeName => Flat; com.eveningoutpost.dexdrip.Extras.BgEstimate => 92.0; com.eveningoutpost.dexdrip.Extras.SourceDesc => G6 Native / G5 Native; com.eveningoutpost.dexdrip.Extras.SourceInfo => G7; com.eveningoutpost.dexdrip.Extras.Collector.NanoStatus =>   ; }Bundle


**Another one:**

10:38:55.902 [main] D/CORE: [DataReceiver.onReceive():38]: onReceive com.eveningoutpost.dexdrip.BgEstimate Bundle{ com.eveningoutpost.dexdrip.Extras.NoiseBlockLevel => 200; com.eveningoutpost.dexdrip.Extras.BgSlope => -6.669134246337811E-6; com.eveningoutpost.dexdrip.Extras.VersionInfo => 2124102913 d14910c-2024.10.29; com.eveningoutpost.dexdrip.Extras.Raw => 0.0; com.eveningoutpost.dexdrip.Extras.Display.Units => mmol; com.eveningoutpost.dexdrip.Extras.Time => 1737797925791; com.eveningoutpost.dexdrip.Extras.CalibrationInfo => -1; com.eveningoutpost.dexdrip.Extras.NsNoiseLevel => null; 

**com.eveningoutpost.dexdrip.Extras.SensorStartedAt => 1737718830699;** 

**The timestamp 1737718830699 corresponds to January 24, 2025, 11:40:30.699 AM according to ChatGPT**

com.eveningoutpost.dexdrip.Extras.Noise => -9999.0; com.eveningoutpost.dexdrip.Extras.SensorBattery => -1; com.eveningoutpost.dexdrip.Extras.BgSlopeName => Flat; com.eveningoutpost.dexdrip.Extras.BgEstimate => 90.0; com.eveningoutpost.dexdrip.Extras.SourceDesc => G6 Native / G5 Native; com.eveningoutpost.dexdrip.Extras.SourceInfo => G7; com.eveningoutpost.dexdrip.Extras.Collector.NanoStatus =>   ; }Bundle


So it's clear unless com.eveningoutpost.dexdrip.Extras.SensorStartedAt is exactly same value as previous entry, a new entry will be created in AAPS and uploaded to NS.

That's why storing the first calculated time for sensorstart in xDrip DB and only sending that until Sender-ID changes, then send new time, for me sounds like a solution that will work without changing anything in AAPS if I understood you correctly.

**Reply by Navid200** [reply]
Any way you can safely test this?  https://github.com/NightscoutFoundation/xDrip/pull/3873  

It is what Milos suggested.

**Reply by jgslade** [reply]
I've built xdrip with #3873 and will test, I've only been getting duplicate start stuff when I started a new sensor. Not sure if having to uninstall and the reinstall will trigger it or not, but if it doesn't I'm on day 5 of my current sensor.

**Reply by Navid200** [reply]
This was a mistake.  Rounding to minutes will reduce the number of instances users experience the start time fluctuating as shown in the first post in this discussion and I am attaching the screenshot here again.
![335822781-b7ff01a6-33c3-48c8-b570-1b0fb83d62fb](https://github.com/user-attachments/assets/ef4c3d57-1a2f-4615-9730-ddf3ad290bd0)

If we round to minutes, there will still be AAPS users who will see this if their G7 start time is close to 30 seconds.
So, yes, there will be less instances.  But, there will still be people who are going to complain.  And what are we going to do then?
I'd rather fix the problem once and be done with it.  

I can open a PR to broadcast what we get from G7 as well as what we are broadcasting now.
But, it would be pointless if AAPS cannot or will not use it.
Can AAPS be modified to do the necessary calculations?

**Reply by Navid200** [reply]
I thought I had a solution.  I was going to round not to 0.  I was going to round either to 0 or 30.  But, I realized that G7 also includes Stelo.  Stelo runs for 15 days.  Over 15 days, the drift could be 20 seconds!  There is no simple solution like a clever rounding that can address that much drift.

**Reply by Navid200** [reply]
Can we take a step back and look at this from a different angle?
Why does anyone care when the sensor started as an absolute value of time?
Why is it not enough to ask "how long ago did the sensor start"?
That value will never fluctuate.

We can broadcast that value.  If AAPS shows that value, instead of the absolute time, the problem will be solved.

**Reply by jgslade** [reply]
Oh man, this is tricky.

Looking through the AAPS code I think this is how AAPS gets the start time in
app.aaps.plugins.source.XdripSourcePlugin.kt

```
 fun getSensorStartTime(bundle: Bundle): Long? {
            val now = dateUtil.now()
            var sensorStartTime: Long? = if (sp.getBoolean(R.string.key_dexcom_log_ns_sensor_change, false)) {
                bundle.getLong(Intents.EXTRA_SENSOR_STARTED_AT, 0)
            } else {
                null
            }
            // check start time validity
            sensorStartTime?.let {
                if (abs(it - now) > T.months(1).msecs() || it > now) sensorStartTime = null
            }
            return sensorStartTime
        }
```

In app.aaps.core.interfaces.receivers.Intents.kt it looks like these are defined:

```
        // xDrip -> AAPS
        const val ACTION_NEW_BG_ESTIMATE = "com.eveningoutpost.dexdrip.BgEstimate"
        const val EXTRA_BG_ESTIMATE = "com.eveningoutpost.dexdrip.Extras.BgEstimate"
        const val EXTRA_BG_SLOPE = "com.eveningoutpost.dexdrip.Extras.BgSlope"
        const val EXTRA_BG_SLOPE_NAME = "com.eveningoutpost.dexdrip.Extras.BgSlopeName"
        const val EXTRA_SENSOR_BATTERY = "com.eveningoutpost.dexdrip.Extras.SensorBattery"
        const val EXTRA_SENSOR_STARTED_AT = "com.eveningoutpost.dexdrip.Extras.SensorStartedAt"
        const val EXTRA_TIMESTAMP = "com.eveningoutpost.dexdrip.Extras.Time"
        const val EXTRA_RAW = "com.eveningoutpost.dexdrip.Extras.Raw"
        const val XDRIP_DATA_SOURCE_DESCRIPTION = "com.eveningoutpost.dexdrip.Extras.SourceDesc"
        const val ACTION_NEW_BG_ESTIMATE_NO_DATA = "com.eveningoutpost.dexdrip.BgEstimateNoData"
        const val NS_EMULATOR = "com.eveningoutpost.dexdrip.NS_EMULATOR"
        const val ACTION_REMOTE_CALIBRATION = "com.eveningoutpost.dexdrip.NewCalibration"
        const val GLIMP_BG = "it.ct.glicemia.ACTION_GLUCOSE_MEASURED"
        const val DEXCOM_BG = "com.dexcom.cgm.EXTERNAL_BROADCAST"
        const val DEXCOM_G7_BG = "com.dexcom.g7.EXTERNAL_BROADCAST"
        const val EVERSENSE_BG = "com.senseonics.AndroidAPSEventSubscriber.BROADCAST"
        const val POCTECH_BG = "com.china.poctech.data"
        const val TOMATO_BG = "com.fanqies.tomatofn.BgEstimate"
```

I have some background in programming, but I don't know java/kotlin so there is probably stuff I'm missing.

**Reply by jgslade** [reply]
I have been running with that patch for about 40 minutes now and have not gotten a connection to my sensor yet. I am getting a no valid timestamp error.

**Comment by Navid200** [comment]
@olorinmaia Does the AAPS code look at the sensor start time and when it changes concludes that there must be a new sensor?
If yes, can a threshold be added to it?   For example, can it be changed to if the sensor start time changes by more than 5 minutes, we conclude that we must have a new sensor?

**Reply by olorinmaia** [reply]
My proposal to fix this. Please test https://github.com/nightscout/AndroidAPS/pull/3804

**Comment by olorinmaia** [comment]
https://github.com/nightscout/AndroidAPS/pull/3804
Has been merged to dev for AAPS so will be included in next master release. If you don't want to wait you can cherry pick it from dev on to your master branch.
